src/cmap/auth/mongodb_oidc/callback_workflow.ts

```diff
@@ -193,79 +132,28 @@ export class CallbackWorkflow implements Workflow {
    * Fetches an access token using either the request or refresh callbacks and
    * puts it in the cache.
    */
-  private async fetchAccessToken(
+  protected async fetchAccessToken(
     connection: Connection,
     credentials: MongoCredentials,
-    serverInfo: IdPServerInfo,
-    reauthenticating: boolean,
-    callbackHash: string,
-    requestCallback: OIDCRequestFunction,
-    refreshCallback?: OIDCRefreshFunction
-  ): Promise<IdPServerResponse> {
-    // Get the token from the cache.
-    const entry = this.cache.getEntry(connection.address, credentials.username, callbackHash);
-    let result;
-    const context: OIDCCallbackContext = { timeoutSeconds: TIMEOUT_S, version: OIDC_VERSION };
-    // Check if there's a token in the cache.
-    if (entry) {
-      // If the cache entry is valid, return the token result.
-      if (entry.isValid() && !reauthenticating) {
-        return entry.tokenResult;
-      }
-      // If the cache entry is not valid, remove it from the cache and first attempt
-      // to use the refresh callback to get a new token. If no refresh callback
-      // exists, then fallback to the request callback.
-      if (refreshCallback) {
-        context.refreshToken = entry.tokenResult.refreshToken;
-        result = await refreshCallback(serverInfo, context);
-      } else {
-        result = await requestCallback(serverInfo, context);
-      }
-    } else {
-      // With no token in the cache we use the request callback.
-      result = await requestCallback(serverInfo, context);
+    callback: OIDCCallbackFunction,
+    idpInfo?: IdPInfo
+  ): Promise<TokenEntry> {
+    const params: OIDCCallbackParams = {
+      timeoutContext: AbortSignal.timeout(TIMEOUT_MS),
+      version: OIDC_VERSION
+    };
+    if (idpInfo) {
+      params.idpInfo = idpInfo;
     }
+    // With no token in the cache we use the request callback.
+    const result = await callback(params);
     // Validate that the result returned by the callback is acceptable. If it is not
     // we must clear the token result from the cache.
```

Comment:
from the spec:

> Call the configured built-in provider integration or the OIDC callback to retrieve a new access token. Wait until it has been at least 100ms since the last callback invocation, to avoid overloading the callback.

Seems like we need to add some rate limiting to ensure we don't overload the callback.  Also, I'd like clarification that we only rate limit the callback workflow and not the machine workflow too.

Reply:
It applies to both